multiraft/multiraft.go

```diff
@@ -550,7 +571,43 @@ func (s *state) start() {
 
 				select {
 				case resp := <-s.writeTask.out:
-					s.handleWriteResponse(resp, writingGroups)
+					callbackCount := s.handleWriteResponse(resp, writingGroups)
+					// Wait till all callback returned and send buffered event if any.
+					if callbackCount > 0 {
+						sendIndex := 0
+						drained := false
+						for {
+							// Check if any pending events can be sent
+							if sendIndex < len(s.pendingEvents) {
+								event = s.pendingEvents[sendIndex]
+								eventsChan = s.Events
+							} else {
+								s.pendingEvents = make([]interface{}, 0)
+								eventsChan = nil
+								drained = true
+							}
+							var callbackChan chan func()
+							// Check if all callback had been returned, if yes, break the looop
+							if callbackCount > 0 {
+								callbackChan = s.callbackChan
+							} else {
+								// Residual pending events will be sent by outer select
+								if !drained {
+									s.pendingEvents = s.pendingEvents[sendIndex:]
+								}
+								break
+							}
+							select {
+							case <-s.stopper.ShouldStop():
+								return
+							case eventsChan <- event:
+								sendIndex++
```

Comment:
If you set `s.pendingEvents = s.pendingEvents[1:]` after each event is sent, I think you can simplify this code by getting rid of the `draining` variable.